### ISSUE TITLE
False positive used-before-assignment when assignment can be an import

### ISSUE THREAD

**Issue by mthuurne** [PyCQA/pylint#3127]
### Steps to reproduce

Run pylint on the following code:

```py
from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from collections import Counter
else:
    Counter = object

def tick(c: Counter, name: str) -> None:
    c[name] += 1
```

While this example is rather pointless, I use similar constructs a lot when I have to import a name only for the purpose of type checking. It helps avoid import cycles between modules.

### Current behavior

pylint reports on the `def tick` line:
```
E0601: Using variable 'Counter' before assignment (used-before-assignment)
```

### Expected behavior

No errors are reported, since `Counter` is guaranteed to be defined.

### pylint --version output

pylint 2.4.0
astroid 2.3.0
Python 3.7.2 (default, Feb 26 2019, 13:02:33) 
[GCC 7.3.1 20180323 [gcc-7-branch revision 258812]]

**Comment by mthuurne** [comment]
That's quick. Thanks!

**Comment by mthuurne** [comment]
It seems the fix in 2.4.1 fixes the case where one type is imported inside a type guard, but when multiple types are imported, there is still an error reported for every type except the first.

For this test case:
```py
from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from collections import Counter
    from collections import OrderedDict
else:
    Counter = object
    OrderedDict = object

def tick(c: Counter, deltas: OrderedDict) -> None:
    for name, delta in deltas.items():
        c[name] += delta
```
pylint reports:
```
E0601: Using variable 'OrderedDict' before assignment (used-before-assignment)
```

**Comment by Deimos** [comment]
I'm still seeing this false positive on a similar case. For me, it's a workaround because mypy doesn't deal with SQLAlchemy's `hybrid_property` decorator correctly, so I do this:

```python
if TYPE_CHECKING:
    from builtins import property as hybrid_property
else:
    from sqlalchemy.ext.hybrid import hybrid_property
```

This makes it so that mypy treats `@hybrid_property` as `@property` and understands the getter/setter methods. However, it's causing a `used-before-assignment` error in pylint 2.4.4.

**Comment by snejus** [comment]
Similar issues here:

```python
from typing import TYPE_CHECKING, Type

if TYPE_CHECKING:
    from django.db.models import Model

def run(sender: Type[Model]):
```

**Comment by mthuurne** [comment]
> Same issue here, pylint `2.5.2`:
> 
> ```python
> from typing import TYPE_CHECKING, Type
> 
> if TYPE_CHECKING:
>     from django.db.models import Model
> 
> def run(sender: Type[Model]):
> ```

I think in that example pylint is correct to report an error, since that will lead to a runtime error when executed and `Model` is undefined when the `def` statement runs.

**Comment by snejus** [comment]
You are entirely correct - sorry for making waves :)

**Comment by afischer-sh** [comment]
It happens again in pyling 4.0.0 - please reopen:

pylint 4.0.0
astroid 4.0.1
Python 3.14.0

**Comment by Pierre-Sassoulas** [comment]
@afischer-sh do you mind providing a reproducer, please ?

**Comment by AlekseiSokolov2TR** [comment]
@Pierre-Sassoulas 

```
$ pylint --version
pylint 4.0.4
astroid 4.0.2
Python 3.12.3 (main, Nov  6 2025, 13:44:16) [GCC 13.3.0]
```

```
# pylint: disable=missing-docstring,too-few-public-methods
import json
from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from mypy_boto3_secretsmanager.client import SecretsManagerClient


class Secret:
    def __init__(self, secret_arn: str, client: SecretsManagerClient):
        self.arn = secret_arn
        self.client = client
```

```
secret.py:10:48: E0601: Using variable 'SecretsManagerClient' before assignment (used-before-assignment)
```

**Comment by Pierre-Sassoulas** [comment]
@AlekseiSokolov2TR  you don't seems to use from future import annotations, pylint still support older versions of python (3.10 currentely)
